[JerryO3] iP

JerryO3 opened the pull request:
# Elias is here to serve  you! :man_in_tuxedo: 

> "It's not enough to be busy, so are the ants. The question is, what are we busy about?" - Henry David Thoreau

Elias is an ongoing project of [Project Duke](https://github.com/se-edu/duke), which is a project template for a greenfield Java project from [SE-EDU](https://se-education.org/docs/templates.html). Elias is a lightweight task list manager that currently supports:

1. **Adding 3 types of Tasks**

   - ***Todos*** for tasks with no deadline or time period 
   `todo <todo name>`
   - ***Deadlines*** for tasks that are associated with a deadline 
   `deadline <deadline name>`
   - ***Events*** for tasks that are associated with a start and end time 
   `event <event name>`
2. **Deleting Tasks**
`delete <task index>`
4. **Marking Tasks as complete/incomplete**
`mark<task index>` / `unmark <task index>`
    - ~~Mark multiple tasks at one go~~ This feature has been deprecated :pensive:
6. **Listing current tasks**
`list`
8. **Searching for tasks by name**
`find <search string>`
10. **Parsing multiple string formats of dates and times:**
      - `dd/mm/yy HHmm`
      - `dd/mm/yy hhmmAM` or `dd/mm/yy hhmmPM`
      - `dd-mm-yy HHmm`
      - `dd-mm-yy hhmmAM` or `dd-mm-yy hhmmPM`
      - `dd/mm/yy, hhmm AM` or `dd/mm/yy, hhmm PM`
12. **Saving and Loading the current tasklist to memory**

Future features may include:
- [x] JavaFX UI (In progress)
       - Light theme background `#ffffff`
       - Dark theme background `#0f3857`
- [ ] Sorting deadlines and events by time (Depending on timeline)

---
### Useful stuff for contributors:

Extensible Command list:
```java
    public Parser(ItemList itemlist) {
        this.itemList = itemlist;
        commands.put("event", new AddCommand());
        commands.put("todo", new AddCommand());
        commands.put("deadline", new AddCommand());
        commands.put("mark", new MarkCommand());
        commands.put("unmark", new UnmarkCommand());
        commands.put("delete", new DeleteCommand());
        commands.put("list", new ListCommand());
        commands.put("find", new FindCommand());
    }
```

Extensible DateTime parsing pattern list:
```java
    private static DateTimeFormatterBuilder[] dtFormats = {
            new DateTimeFormatterBuilder().append(DateTimeFormatter.ofLocalizedDateTime(FormatStyle.SHORT)),
            new DateTimeFormatterBuilder().appendPattern("dd/MM/yy-HHmm"),
            new DateTimeFormatterBuilder().appendPattern("dd/MM/yy HHmm"),
            new DateTimeFormatterBuilder().appendPattern("dd-MM-yy-HHmm"),
            new DateTimeFormatterBuilder().appendPattern("dd-MM-yy HHmm"),
            new DateTimeFormatterBuilder().appendPattern("dd/MM/yy-hh:mma"),
            new DateTimeFormatterBuilder().appendPattern("dd/MM/yy hh:mma"),
            new DateTimeFormatterBuilder().appendPattern("dd-MM-yy-hh:mma"),
            new DateTimeFormatterBuilder().appendPattern("dd-MM-yy hh:mma")
    };
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

JerryO3 committed:
renamed, added greet and exit

JerryO3 committed:
added Echo file

JerryO3 committed:
modified Duke and Greet to perform echo function

JerryO3 committed:
modified Duke so that it no longer exits on the first command

JerryO3 committed:
modified Greet and Echo for indent and horizontal rule specification

JerryO3 committed:
added itemlist, not formatted yet

JerryO3 committed:
added standard formatting for string outputs

JerryO3 committed:
Modified Itemlist: changed implementation to list of String

JerryO3 committed:
added mark and unmark functionality

JerryO3 committed:
created item class, and made ItemList a class whose instance contains a list of items

JerryO3 committed:
fixed formatting for mark/unmark

JerryO3 committed:
modified most files to new subclass requirements, adding todo, deadline and events

JerryO3 committed:
automated testing some testcases

JerryO3 committed:
added custom exceptions for when /to is before /from, and for 0th index task

JerryO3 committed:
added exceptions to handle typos and parsing errors for Event and Deadline

JerryO3 committed:
increased test scope

JerryO3 committed:
added more testcases into input file

JerryO3 committed:
added delete functionality

JerryO3 committed:
new storage file

JerryO3 committed:
testing out serializable

JerryO3 committed:
Converted Item to interface so that it is serializable

JerryO3 committed:
fixed bug where program would exit without an existing save file

JerryO3 committed:
Merge branch 'branch-Level-7'

JerryO3 committed:
replaced time values using datetime type, added Parser class to parse different datetime formats

JerryO3 committed:
Merge branch 'branch-Level-8'

JerryO3 committed:
Add Command.java interface and implementing classes AddCommand, DeleteCommand etc

JerryO3 committed:
Packaged code into duke,  duke.command and duke.item packages

JerryO3 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

JerryO3 committed:
add gradle support

JerryO3 committed:
test Event constructor, test Parser parseDTString method

JerryO3 committed:
update gradle build

JerryO3 committed:
gradle: modify gradle mainClass.set path, Storage: add backup path

JerryO3 committed:
fix some style to conform to style guide

JerryO3 committed:
Add checkstyle files to scan for violations

JerryO3 committed:
fix style to conform to style guide

JerryO3 committed:
Merge branch 'master' into branch-A-CodingStandard

JerryO3 committed:
Merge branch 'master' into branch-Level-9

JerryO3 committed:
Merge branch 'master' into branch-A-Javadoc

JerryO3 committed:
restore data/duke.txt

JerryO3 committed:
add FindCommand class ItemList: add find function Parser: add FindCommand into commands

JerryO3 committed:
Added java documentation for command package

JerryO3 committed:
Add Java documentation for item package

JerryO3 committed:
Add Java documentation for CustomExceptions Fixed 0-index bug for delete, mark and unmark command

JerryO3 committed:
Add Java documentation for Duke, ItemList, Parser, Storage and UI classes

JerryO3 committed:
Merge branch 'branch-A-Javadoc'

JerryO3 committed:
Merge branch 'branch-Level-9'

JerryO3 committed:
Merge branch 'branch-A-CodingStandard'

JerryO3 committed:
Added Java documentation for find method in ItemList and for FindException. Removed unused import in Todo.

JerryO3 committed:
Add Java documentation for FindCommand and its constructor. Fixed style in Command

cheahTJ commented on src/main/java/duke/CustomExceptions.java:
```diff
@@ -0,0 +1,129 @@
+package duke;
+
```
Despite the many exceptions classes created in this class, I really like how how the methods are named and that makes it so intuitive to understand them even without reading the comments.

cheahTJ commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+ * unless no such directory exists, where it is added to the root.
+ */
+public class Duke {
+    private final Storage storage;
```
Good use of variable name, similar to the type of the class they are referring to. This makes the tracking of the logic in the code so much easier!

cheahTJ commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+    public static void main(String[] args) {
+        Duke elias = new Duke("./data/duke.txt");
+        System.out.println(elias.ui.getLogo());
+        System.out.println(elias.ui.format(elias.ui.getGreet()));
```
I really like how the method here is short and easy to understand but do you think it might perhaps be better to rename the method to just `greet()` ? The rationale for this is that it might make your code even more readable as with the proposed change, it would be flow-y to read it as 'elias greet'.

cheahTJ commented on src/main/java/duke/ItemList.java:
```diff
@@ -0,0 +1,131 @@
+     * @throws CustomExceptions.NoSuchIndexException when no such index
+     *                                               exists in the ItemList
+     */
+    public String removeItem(int index) throws CustomExceptions {
```
Perhaps to make it even more clear that you are removing the item from the list of items when calling this method from other parts of your code base, would it be better to rename this method as `removeItemFromlist(int index) `?

cheahTJ commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return an ItemList object that has been deserialized from duke.txt
+     */
+    public ItemList readFromFile() {
```
I really appreciate how intuitive this method is but do you think it would be good if you could phrase it as `bootingUp()` or `initialiseItemList()`?  The rationale for this proposed change is because since this method will be called as the first touch point between your local file storage and your ItemList object when you want to load the content from the file to the ItemList object, would it paint a clearer picture if you were to use the proposed method name?

cheahTJ commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,23 @@
+     * @throws CustomExceptions which encompasses all custom exceptions in this
+     *                          project
+     */
+    String execute(String command, String[] info, ItemList itemList) throws CustomExceptions;
```
Very good use of verb 'execute' to denote the start of taking in the user's input and makes sense of it :)

cheahTJ reviewed (commented):
LGTM, just some small pointers to consider in my comments. Thank you!

JerryO3 committed:
Modify accepted DateTime string formats

JerryO3 committed:
Add GUI using JavaFX

JerryO3 committed:
Add comments

JerryO3 committed:
Merge branch 'branch-Level-10'

johnyoozhengxian commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,102 @@
+            new DateTimeFormatterBuilder().appendPattern("dd/MM/yy hh:mma"),
+            new DateTimeFormatterBuilder().appendPattern("dd-MM-yy-hh:mma"),
+            new DateTimeFormatterBuilder().appendPattern("dd-MM-yy hh:mma"),
+            new DateTimeFormatterBuilder().appendPattern("d/M/y HHmm")
```
good number of datetime formats

johnyoozhengxian reviewed (commented):
I think your code is very good and clear, also with good use of OOP. Not much to complain about. Keep it up!

johnyoozhengxian commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,47 @@
+    private String greet = "Hello! I'm ELIAS\n"
+            + "     What can I do for you?";
+
+    private String logo = "______ _ _\n"
```
Clever creation of the word Elias using text.

johnyoozhengxian commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,73 @@
+            } else {
+                return itemList.addToDo(info);
+            }
+        } else if (info[0].equals("deadline")) {
```
Clear check for the inputs. Code is easy to understand.

johnyoozhengxian commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,46 @@
+     *                                               passed.
+     */
+    @Override
+    public String execute(String command, String[] info, ItemList itemList) throws CustomExceptions {
```
Good coding style by using self made exceptions and not just a println.

johnyoozhengxian commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,35 @@
+ * current ItemList that contains a given search string, parsed
+ * from command.
+ */
+public class FindCommand implements Command {
```
Good use of OOP.

johnyoozhengxian reviewed (commented):
Nothing to complain about. Sorry for reviewing a second time because my first review did not have enough comments. Very good job.

JerryO3 committed:
Remove main() in Duke, allow application close on "bye" command

main() function in Duke class is no longer used due to changes from
implementing JavaFX. In addition, some responses do not render
properly on the new GUI.

Remove the main() function, add the greet and logo methods in the start
method in Duke. Also add the Platform exit() function to the function
handling user input.

JerryO3 committed:
Implement JavaFXML for MainWindow and DialogueBox controllers

As instructed in the increments for week 4, also to better understand
how fxml works.

Let's add MainWindow and DialogueBox fxml files. Also add resizing
functionality to the labels in dialogue box.

This allows us to directly modify the GUI using screenbuilder, and for
long text to be fully captured in labels

JerryO3 committed:
Style(GUI): Modify fxml files using scene builder to improve GUI

Base GUI is ugly

Let's add some rounded corners, padding and colors

Neutral colors are pleasing to look at

JerryO3 committed:
Style(Styleguide): Add javadocs and change arrangement of fields

JerryO3 committed:
Add assertions to check for nullable entities\n\nEnsure that invalid null state does\n not occur during runtime\n\nAdd assertions to check if null\n\nAs required by specification

JerryO3 committed:
Improve code quality: use default branch in AddCommand

JerryO3 committed:
Refactor AddCommand to switch statement to reduce arrowhead code

JerryO3 committed:
Refactor Add Command execute method

The execute method was more than 30 LOC because it had a common
nested if-condition

Let's extract the nested if-condition to shorten the code. Also shift
error handling away from execute directly to Parser parse and the
respective task classes

This helps to shorten the method considerably, and also better
delineate responsibility of error handling to the correct classes.

JerryO3 committed:
Rearrange code blocks in Deadline constructor

JerryO3 committed:
Remove unused zeroIndexException

JerryO3 committed:
Fix typo on Javadoc for FindCommand

JerryO3 committed:
Add final keyword to fields that should not be modified

JerryO3 committed:
Remove redundant initialized values in Duke class

JerryO3 committed:
Refactor code using inline method to improve readability

JerryO3 committed:
Merge pull request #2 from JerryO3/branch-A-CodeQuality

Improve code quality

JerryO3 committed:
Deconflict merge master into branch-A-Assertions

JerryO3 committed:
Merge pull request #3 from JerryO3/branch-A-Assertions

Branch a assertions

JerryO3 committed:
Add sort command as new class, sort function in item list

JerryO3 committed:
Add sortcommand into list of executable commands in parser

JerryO3 committed:
Add get-time-to-sort-by method so that item list can be sorted by time

JerryO3 committed:
Configure gradle for fat jar

JerryO3 committed:
Add one more local date time pattern

JerryO3 committed:
Edit readme to contain feature list

JerryO3 committed:
Update README at root dir

JerryO3 committed:
Update READMEs to include bye command

JerryO3 committed:
Change ui image filename

JerryO3 committed:
Rename UI.png to Ui.png
